### DISCUSSION TITLE
[Feature] Image accent color determiner

### DISCUSSION THREAD

**Discussion by Avid29** [CommunityToolkit/Labs-Windows#254] [Ideas]
### Describe the problem

A fairly common feature in my opinion is having the accent color of an app be related to an image shown. 
There is not any massively leading library for this, and while Color Thief .NET exists, it is not currently being updated.

### Describe the solution

I would like to add a service to the library that determines the accent color of images for easy use.

### Alternatives

Color Thief .NET (outdated)

### Additional info

I initially wrote some code for this in Strix Music v2, though I have significantly refined it since. I'm also interested in writing a version using [MeanShift++ ](https://openaccess.thecvf.com/content/CVPR2021/papers/Jang_MeanShift_Extremely_Fast_Mode-Seeking_With_Applications_to_Segmentation_and_Object_CVPR_2021_paper.pdf) in order to speed the process up.

I'm not certain how the service would be used. I believe it would ideally be done somehow using attached properties to set a resource that could be looked up, though I'd like a discussion with someone more experienced in attached properties to discover what that may look like.

### Help us help you

Yes, I'd like to be assigned to work on this item.

**Comment by ghost** [comment]
Hello, 'Avid29! Thanks for submitting a new feature request. I've automatically added a vote 👍 reaction to help get things started. Other community members can vote to help us prioritize this feature in the future!

**Comment by Avid29** [comment]
I of course expect this to be a toolkit labs thing for now

**Comment by Arlodotexe** [comment]
Transferred to Labs 🙂 

To start out, you probably want to keep it simple: something that can take a stream and spit out a color.

**Comment by Avid29** [comment]
I have a proof of concept working with the following issues:
- Naïve cluster used (No constraints for saturation or value applied)
- Pixels are poorly sampled (basically just grabs the color of the top row)
- Currently only works in .NET 6
- Runs a little slow (50ms)

![image](https://github.com/CommunityToolkit/Labs-Windows/assets/9384045/5625f216-0c69-477f-92cd-c882c9419528)

This is how pixel data is grabbed, which is also where NET 6 is required.

```cs
var stream = await RandomAccessStreamReference.CreateFromUri(imgSrc.UriSource).OpenReadAsync();
var decoder = await BitmapDecoder.CreateAsync(stream);
var pixelData = await decoder.GetPixelDataAsync();
var bytes = pixelData.DetachPixelData();
```
``RandomAccessStreamReference.CreateFromUri`` is not available in net standard.

Next steps:

- Introduce cluster post-processing
- Randomize pixel sampling
- Find alternate pixel decoding API
- Optimize Clustering
  - Centroid deferred computation
  - Iteration cutoff
  - Replace KMeans with MeanShift++ (potentially)
